pkg/objstore/azure/azure.go

```diff
@@ -0,0 +1,292 @@
+package azure
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+
+	blob "github.com/Azure/azure-storage-blob-go/2018-03-28/azblob"
+	"github.com/improbable-eng/thanos/pkg/objstore"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	yaml "gopkg.in/yaml.v2"
+)
+
+const (
+	opObjectsList  = "ListBucket"
+	opObjectInsert = "PutObject"
+	opObjectGet    = "GetObject"
+	opObjectHead   = "HeadObject"
+	opObjectDelete = "DeleteObject"
+)
+
+// Config Azure storage configuration.
+type Config struct {
+	StorageAccountName string `yaml:"storage-account"`
+	StorageAccountKey  string `yaml:"storage-account-key"`
+	ContainerName      string `yaml:"container"`
+}
+
+// Bucket implements the store.Bucket interface against s3-compatible APIs.
+type Bucket struct {
+	logger       log.Logger
+	containerURL blob.ContainerURL
+	config       *Config
+	opsTotal     *prometheus.CounterVec
+	closer       blobReadCloser
+}
+
+// Validate checks to see if any of the s3 config options are set.
+func (conf *Config) Validate() error {
+	if conf.StorageAccountName == "" ||
+		conf.StorageAccountKey == "" {
+		return errors.New("invalid Azure storage configuration")
+	}
+	return nil
+}
+
+// NewBucket returns a new Bucket using the provided Azure config.
+func NewBucket(logger log.Logger, azureConfig []byte, reg prometheus.Registerer, component string) (*Bucket, error) {
+
+	level.Debug(logger).Log("msg", "Creating new Azure bucket connection", "component", component)
+
+	var conf Config
+	if err := yaml.Unmarshal(azureConfig, &conf); err != nil {
+		return nil, err
+	}
+
+	err := conf.Validate()
+	if err != nil {
+		return nil, err
+	}
+
+	ctx := context.Background()
+
+	container, err := createContainer(ctx, conf.StorageAccountName, conf.StorageAccountKey, conf.ContainerName)
```

Comment:
I would really put all of this huge `if` and `else` cases into `createContainer` function. But up to you

Reply:
I would keep it here, because I don't really want to put whole logic into helper function which does only one little thing. I also need to pass logger and handle errors in both places etc..